### PR TITLE
215 consider when version is not provided

### PR DESCRIPTION
**Pull request by jacopo-exact**
fill package version information in source if needed.

add test

ref #215

### CODE REVIEW

File: tests/test_task_api.py [hunk 1]
```diff
@@ -22,6 +27,41 @@ async def test_task_get_list(db, client, task_factory, MockCurrentUser):
...
+        task_pkg=task_pkg, user="test_bg_collection"
+    )
+    collection_status = TaskCollectStatus(
+        status="pending",
+        venv_path=venv_path,
+        package=task_pkg.package
+        # **task_pkg.dict(),
+    )
+    # replacing with path because of non-serializable Path
+    collection_status_dict = collection_status.dict()
```
Comment (tcompa) [inline tests/test_task_api.py:47]: Can we use `sanitised_dict` here?

File: tests/test_task_api.py [hunk 2]
```diff
@@ -22,6 +27,41 @@ async def test_task_get_list(db, client, task_factory, MockCurrentUser):
...
+    """
+    task_pkg = _TaskCollectPip(package=dummy_task_package.as_posix())
+    venv_path = create_package_dir_pip(
+        task_pkg=task_pkg, user="test_bg_collection"
+    )
+    collection_status = TaskCollectStatus(
+        status="pending",
+        venv_path=venv_path,
+        package=task_pkg.package
+        # **task_pkg.dict(),
```
Comment (tcompa) [inline tests/test_task_api.py:44]: Is this comment relevant?

File: fractal_server/app/api/v1/task.py [hunk 3]
```diff
@@ -94,51 +152,67 @@ async def collect_tasks_pip(
...
-    settings = Inject(get_settings)
 
-    return TaskCollectResult(
-        **task_pkg.dict(),
-        venv_path=venv_path.relative_to(settings.FRACTAL_ROOT),
-        info=(
-            "Collecting tasks in the background. "
-            "GET /task/collect/{venv_path} to query collection status"
-        ),
+    logger.info("collection endpiont: returning state")
```
Comment (tcompa) [inline fractal_server/app/api/v1/task.py:179]: typo

**Review by tcompa — APPROVED** [review]

**Comment by tcompa** [comment]
One last comment, for my understanding.
I added `import time; time.sleep(10)` within `_background_collect_pip` (for testing), and then ran `poetry run pytest tests/test_task_api.py::test_collection_api`. I was expecting this block to fail
```python
       res = await client.post(
            f"{PREFIX}/collect/pip/?public=false", json=task_collection
        )
        debug(res.json())
        assert res.status_code == 201
        assert res.json()["data"]["status"] == "pending"

        state = res.json()
        data = state["data"]
        assert "fractal_tasks_dummy" in data["venv_path"]
        venv_path = Path(data["venv_path"])

        res = await client.get(f"{PREFIX}/collect/{state['id']}")
        debug(res.json())
        assert res.status_code == 200
        status = res.json()
        data = status["data"]

        assert data["status"] == "OK"
```
because the background task should have taken longer than the time to get to the OK assert statement.

What I observed is that the test passed, but simply took 10 seconds longer. I don't understand which call was blocking, and why the 10 seconds were not being waited in background.

(nothing to change, most likely, but I just want to understand better)

File: tests/test_task_api.py (hunk 1 above)
Comment (jacopo-exact) [reply tests/test_task_api.py:47]: yes, thanks

File: tests/test_task_api.py (hunk 2 above)
Comment (jacopo-exact) [reply tests/test_task_api.py:44]: no, thanks

File: tests/test_task_api.py [hunk 4]
```diff
@@ -46,16 +86,18 @@ async def test_collection_api(client, dummy_task_package, MockCurrentUser):
...
+        data = state["data"]
         assert "fractal_tasks_dummy" in data["venv_path"]
         venv_path = Path(data["venv_path"])
 
-        res = await client.get(f"{PREFIX}/collect/{venv_path}")
+        res = await client.get(f"{PREFIX}/collect/{state['id']}")
         debug(res.json())
         assert res.status_code == 200
-        data = res.json()
+        status = res.json()
```
[the remark this answers is not in the document]
Comment (jacopo-exact) [reply tests/test_task_api.py:99]: ok

File: tests/test_task_api.py [hunk 5]
```diff
@@ -67,14 +109,39 @@ async def test_collection_api(client, dummy_task_package, MockCurrentUser):
         assert "dummy parallel" in task_names
 
         # using verbose option
-        res = await client.get(f"{PREFIX}/collect/{venv_path}?verbose=true")
+        res = await client.get(f"{PREFIX}/collect/{state['id']}?verbose=true")
         debug(res.json())
+        status = res.json()
```
[the remark this answers is not in the document]
Comment (jacopo-exact) [reply tests/test_task_api.py:114]: ok

**Comment by jacopo-exact** [comment]
> One last comment, for my understanding. I added `import time; time.sleep(10)` within `_background_collect_pip` (for testing), and then ran `poetry run pytest tests/test_task_api.py::test_collection_api`. I was expecting this block to fail
> 
> ```python
>        res = await client.post(
>             f"{PREFIX}/collect/pip/?public=false", json=task_collection
>         )
>         assert res.status_code == 201
>         assert res.json()["data"]["status"] == "pending"
> 
>         ...
> 
>         res = await client.get(f"{PREFIX}/collect/{state['id']}")
>         assert res.status_code == 200
>         status = res.json()
>         data = status["data"]
> 
>         assert data["status"] == "OK"
> ```
> 
> because the background task should have taken longer than the time to get to the OK assert statement.
> 
> What I observed is that the test passed, but simply took 10 seconds longer. I don't understand which call was blocking, and why the 10 seconds were not being waited in background.
> 
> (nothing to change, most likely, but I just want to understand better)

Most likely it did not fail because you used the synchronous sleep and the test runs in an event loop within a single thread. `_background_collect_pip` acquired the loop and could not yield back because of the synchronous call.

I would expect that if you repeted your experiment with an `await asyncio.sleep(10)` the test would fail.

**Pull request merged**